### ISSUE TITLE
Connection Name Error

### ISSUE THREAD

**Issue by geekhack** [balderdashy/sails-mysql#158]
I am getting this error after updating sails-mysql:

```
etarmacking/node_modules/sails-mysql/lib/adapter.js:59
      if (evName === 'error') { throw data; }
                                      ^
AdapterError: Invalid connection name specified
    at getConnectionObject (/home/cis/Desktop/finalgit/etarmacking/node_modules/sails-mysql/lib/adapter.js:1205:35)
    at spawnConnection (/home/cis/Desktop/finalgit/etarmacking/node_modules/sails-mysql/lib/adapter.js:1120:7)
    at Object.module.exports.adapter.describe (/home/cis/Desktop/finalgit/etarmacking/node_modules/sails-mysql/lib/adapter.js:116:16)
    at module.exports.describe (/home/cis/Desktop/finalgit/etarmacking/node_modules/sails/node_modules/waterline/lib/waterline/adapter/ddl.js:27:18)
    at module.exports.migrateAlter (/home/cis/Desktop/finalgit/etarmacking/node_modules/sails/node_modules/waterline/lib/waterline/adapter/sync.js:28:10)
    at Query.sync (/home/cis/Desktop/finalgit/etarmacking/node_modules/sails/node_modules/waterline/lib/waterline/query/index.js:44:62)
    at /home/cis/Desktop/finalgit/etarmacking/node_modules/sails/node_modules/waterline/lib/waterline/collection/index.js:53:10
    at /home/cis/Desktop/finalgit/etarmacking/node_modules/sails/node_modules/waterline/lib/waterline/adapter/setupTeardown.js:36:7
    at /home/cis/Desktop/finalgit/etarmacking/node_modules/sails/node_modules/async/lib/async.js:116:25
    at /home/cis/Desktop/finalgit/etarmacking/node_modules/sails/node_modules/async/lib/async.js:24:16
    at register (/home/cis/Desktop/finalgit/etarmacking/node_modules/sails/node_modules/waterline/lib/waterline/adapter/setupTeardown.js:31:7)
    at /home/cis/Desktop/finalgit/etarmacking/node_modules/sails/node_modules/async/lib/async.js:108:13
    at Array.forEach (native)
    at _each (/home/cis/Desktop/finalgit/etarmacking/node_modules/sails/node_modules/async/lib/async.js:32:24)
    at Object.async.each (/home/cis/Desktop/finalgit/etarmacking/node_modules/sails/node_modules/async/lib/async.js:107:9)
    at module.exports.registerCollection (/home/cis/Desktop/finalgit/etarmacking/node_modules/sails/node_modules/waterline/lib/waterline/adapter/setupTeardown.js:34:11)
    at module.exports (/home/cis/Desktop/finalgit/etarmacking/node_modules/sails/node_modules/waterline/lib/waterline/collection/index.js:50:17)
    at new child (/home/cis/Desktop/finalgit/etarmacking/node_modules/sails/node_modules/waterline/lib/waterline/utils/extend.js:17:39)
    at instantiateCollection (/home/cis/Desktop/finalgit/etarmacking/node_modules/sails/lib/hooks/orm/index.js:174:25)
    at /home/cis/Desktop/finalgit/etarmacking/node_modules/sails/node_modules/async/lib/async.js:108:13
    at Array.forEach (native)
    at _each (/home/cis/Desktop/finalgit/etarmacking/node_modules/sails/node_modules/async/lib/async.js:32:24)
    at Object.async.each (/home/cis/Desktop/finalgit/etarmacking/node_modules/sails/node_modules/async/lib/async.js:107:9)
    at Hook.startORM (/home/cis/Desktop/finalgit/etarmacking/node_modules/sails/lib/hooks/orm/index.js:164:10)
    at Array.bound [as 1] (/home/cis/Desktop/finalgit/etarmacking/node_modules/sails/node_modules/lodash/dist/lodash.js:729:21)
    at listener (/home/cis/Desktop/finalgit/etarmacking/node_modules/sails/node_modules/async/lib/async.js:462:46)
    at /home/cis/Desktop/finalgit/etarmacking/node_modules/sails/node_modules/async/lib/async.js:416:17
    at Array.forEach (native)
    at _each (/home/cis/Desktop/finalgit/etarmacking/node_modules/sails/node_modules/async/lib/async.js:32:24)
    at Object.taskComplete (/home/cis/Desktop/finalgit/etarmacking/node_modules/sails/node_modules/async/lib/async.js:415:13)
    at processImmediate [as _immediateCallback] (timers.js:330:15)
```

**Comment by IranPhpMaster** [comment]
when I chnage the .exec() in models with .then() it start working fine !!!

**Comment by sailsbot** [comment]
Thanks for posting, @geekhack.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by Ryanc1256** [comment]
@particlebanana hey, i just got this too, clearly not fixed :P

**Comment by Ryanc1256** [comment]
This is a stack trace from my error...
```
error: Sending 500 ("Server Error") POST /v1/auth/login params:{"username":"username","password":"password!","access_token":"ytfeLLB1SnG3pykscyaDF9"} response: 
 AdapterError: Invalid connection name specified
    at getConnectionObject (/app/node_modules/sails-mysql/lib/adapter.js:1182:35)
    at spawnConnection (/app/node_modules/sails-mysql/lib/adapter.js:1097:7)
    at Object.module.exports.adapter.find (/app/node_modules/sails-mysql/lib/adapter.js:801:16)
    at module.exports.find (/app/node_modules/sails/node_modules/waterline/lib/waterline/adapter/dql.js:120:13)
    at module.exports.findOne (/app/node_modules/sails/node_modules/waterline/lib/waterline/adapter/dql.js:163:10)
    at _runOperation (/app/node_modules/sails/node_modules/waterline/lib/waterline/query/finders/operations.js:408:29)
    at run (/app/node_modules/sails/node_modules/waterline/lib/waterline/query/finders/operations.js:69:8)
    at bound.module.exports.findOne (/app/node_modules/sails/node_modules/waterline/lib/waterline/query/finders/basic.js:78:16)
    at bound [as findOne] (/app/node_modules/sails/node_modules/lodash/dist/lodash.js:729:21)
    at Deferred.exec (/app/node_modules/sails/node_modules/waterline/lib/waterline/query/deferred.js:501:16)
    at Strategy._verify (/app/api/services/passport.js:29:7)
    at Strategy.authenticate (/app/node_modules/passport-local/lib/strategy.js:90:12)
    at attempt (/app/node_modules/passport/lib/middleware/authenticate.js:341:16)
    at authenticate (/app/node_modules/passport/lib/middleware/authenticate.js:342:7)
    at Object.AuthController.login (/app/api/controllers/AuthController.js:119:5)
    at bound (/app/node_modules/sails/node_modules/lodash/dist/lodash.js:729:21)
    at routeTargetFnWrapper (/app/node_modules/sails/lib/router/bind.js:179:5)
    at callbacks (/app/node_modules/sails/node_modules/express/lib/router/index.js:164:37)
    at param (/app/node_modules/sails/node_modules/express/lib/router/index.js:138:11)
    at pass (/app/node_modules/sails/node_modules/express/lib/router/index.js:145:5)
    at nextRoute (/app/node_modules/sails/node_modules/express/lib/router/index.js:100:7)
    at callbacks (/app/node_modules/sails/node_modules/express/lib/router/index.js:167:11)
```

**Comment by devinivy** [comment]
What is the name of the connection for the collection on which you're trying to `find`?  Mind posting a collection definition and connections config?

**Comment by Ryanc1256** [comment]
It's just called sails-mysql, and the collection is called User

So this is my connection config...
```
module.exports = {
    adapter: 'sails-mysql',
    host: "localhost",
    user: "root",
    password: "",
    database: "gamefroot"
};
```


and here is the collection definition..
```
module.exports = {
	SUBSCRIPTION_NONE: "none",
	SUBSCRIPTION_PRO: "pro",
	SUBSCRIPTION_ADMIN: "admin",

	types: {
		username: function(username){
			return usernamer.isValid(username);
		}
	},

  	attributes: {

  		id: {
          type: 'integer',
          primaryKey: true,
          autoIncrement: true
        },

  		emails: {
	  		collection: "email",
	  		via: "owner",
	  	},

	  	username: {
	  		type: "string",
	  		unique: true,
	  		required: true,
	  		username: true
	  	},

	  	password: {
	  		type: "string",
	  		required: true,
	  		protected: true,
	  		minLength: 6
	  	},

	  	subscriptionType: {
	  		type: "string",
	  		enum: ["none","pro", "admin"],
	  		defaultsTo: "none"
	  	},

	  	firstName: {
	  		type: "string",
	  		defaultsTo: ""
	  	},

	  	lastName: {
	  		type: "string",
	  		defaultsTo: ""
	  	},

	  	passports: {
	  		collection: "passport",
	  		via: "user"
	  	},

	  	projects: {
	  		collection: "project",
	  		via: "owner"
	  	},

	  	packs: {
	  		collection: "pack",
	  		via: "owner"
	  	},

	  	assets: {
	  		collection: "asset",
	  		via: "owner"
	  	},

	  	purchases: {
	  		collection: "pack",
	  		via: "purchasedBy"
	  	},

	  	/**
	  	 * Gets the storage use of this user
	  	 * @method getStorageUse
	  	 * @return {Promise}
	  	 */
	  	getStorageUse: function(){
	  		return User.getStorageUse(this.id, this.subscriptionType);
	  	}

  	}
};
```

**Comment by devinivy** [comment]
sails-mysql is the name of the adapter.  But what's the name of the connection?  Based upon the file you sent, I assume you're using waterline with sails?  Maybe these links will help.  Notice the connections are named by the keys in the sails connections config.

http://sailsjs.org/documentation/anatomy/my-app/config/connections-js
http://sailsjs.org/documentation/anatomy/my-app/config/models-js

**Comment by RaniSputnik** [comment]
Hi all, I'm working with @Ryanc1256 on this project and I have more information about the issue.

You're correct, we're using Waterline with Sails. We create a connection for each of our three environments using the following;
```
var environments = ["development","staging","production"];
environments.forEach(function(env){
	try {
		var connectionName = env+'SQL';
		Connections[connectionName] = require('./sql/'+env).sql;
	} catch(e) {
		console.warn("Missing SQL configuration for environment",env);
	}
});

module.exports.connections = Connections;
```
Then we have three SQL configuration files in the format Ryan posted earlier;
```
sql /
    development.js
    production.js
    staging.js
```
And in our environment config we specify the connection we require;
```
module.exports = {
    ...
    models: {
        connection: 'productionSQL',
        migrate: 'safe'
    }
}
```
The configuration works correctly for some time, but then - seemingly randomly - we get the following two errors.
```
AdapterError: Invalid connection name specified
    at getConnectionObject (/app/node_modules/sails-mysql/lib/adapter.js:1182:35)
    at spawnConnection (/app/node_modules/sails-mysql/lib/adapter.js:1097:7)
    at Object.module.exports.adapter.find (/app/node_modules/sails-mysql/lib/adapter.js:801:16)
    at module.exports.find (/app/node_modules/sails/node_modules/waterline/lib/waterline/adapter/dql.js:120:13)
    at module.exports.findOne (/app/node_modules/sails/node_modules/waterline/lib/waterline/adapter/dql.js:163:10)
    at _runOperation (/app/node_modules/sails/node_modules/waterline/lib/waterline/query/finders/operations.js:408:29)
    at run (/app/node_modules/sails/node_modules/waterline/lib/waterline/query/finders/operations.js:69:8)
    at bound.module.exports.findOne (/app/node_modules/sails/node_modules/waterline/lib/waterline/query/finders/basic.js:78:16)
    at bound [as findOne] (/app/node_modules/sails/node_modules/lodash/dist/lodash.js:729:21)
    at Deferred.exec (/app/node_modules/sails/node_modules/waterline/lib/waterline/query/deferred.js:501:16)
    at tryCatcher (/app/node_modules/sails/node_modules/waterline/node_modules/bluebird/js/main/util.js:26:23)
    at ret (eval at <anonymous> (/app/node_modules/sails/node_modules/waterline/node_modules/bluebird/js/main/promisify.js:163:12), <anonymous>:13:39)
    at Deferred.toPromise (/app/node_modules/sails/node_modules/waterline/lib/waterline/query/deferred.js:510:61)
    at Deferred.then (/app/node_modules/sails/node_modules/waterline/lib/waterline/query/deferred.js:521:15)
    at Strategy._verify (/app/api/services/passport.js:31:7)
    at Strategy.authenticate (/app/node_modules/passport-local/lib/strategy.js:90:12)
    at attempt (/app/node_modules/passport/lib/middleware/authenticate.js:341:16)
    at authenticate (/app/node_modules/passport/lib/middleware/authenticate.js:342:7)
    at Object.AuthController.login (/app/api/controllers/AuthController.js:119:5)
    at bound (/app/node_modules/sails/node_modules/lodash/dist/lodash.js:729:21)
    at routeTargetFnWrapper (/app/node_modules/sails/lib/router/bind.js:179:5)
    at callbacks (/app/node_modules/sails/node_modules/express/lib/router/index.js:164:37)
```
```
Error (E_UNKNOWN) :: Encountered an unexpected error : 
Could not connect to MySQL: Error: Pool is closed.
    at afterwards (/app/node_modules/sails-mysql/lib/connections/spawn.js:72:13)
    at /app/node_modules/sails-mysql/lib/connections/spawn.js:40:7
    at process._tickDomainCallback (node.js:381:11)
```
We are running on Google Cloud SQL, any thoughts?

**Comment by devinivy** [comment]
It seems like sails must be tearing down the connection.  As far as I know, a teardown is the only way the connection name could become unrecognized, and the adapter never calls this itself.  Do you ever hotload collections or otherwise use `reload` in the sails ORM hook?

**Comment by Ryanc1256** [comment]
Nope, we don't, would it happen if a mysql disconnected?

**Comment by devinivy** [comment]
Even if that were the case, I don't _think_ the adapter would unregister the connection (and forget its name).

**Comment by shanemgrey** [comment]
I'm getting this issue also.  Sails loads everything on lift, but when I change a controller and save, it will sometimes crash the server with this:

    error: Failed to reinitialize ORM.
    error: Error (E_UNKNOWN) :: Encountered an unexpected error
    AdapterError: Invalid connection name specified
        at spawnConnection (/Users/shanegrey/Documents/Code/rmsApi/node_modules/sails-postgresql/lib/adapter.js:1036:43)
        at Object.module.exports.adapter.describe (/Users/shanegrey/Documents/Code/rmsApi/node_modules/sails-postgresql/lib/adapter.js:146:7)
        at module.exports.describe (/usr/local/lib/node_modules/sails/node_modules/waterline/lib/waterline/adapter/ddl/index.js:80:13)
        at module.exports.define (/usr/local/lib/node_modules/sails/node_modules/waterline/lib/waterline/adapter/ddl/index.js:43:10)
        at afterDrop (/usr/local/lib/node_modules/sails/node_modules/waterline/lib/waterline/adapter/sync/strategies/drop.js:35:10)
        at wrapper (/usr/local/lib/node_modules/sails/node_modules/waterline/node_modules/lodash/index.js:3602:19)
        at applyInOriginalCtx (/usr/local/lib/node_modules/sails/node_modules/waterline/lib/waterline/utils/normalize.js:421:80)
        at wrappedCallback (/usr/local/lib/node_modules/sails/node_modules/waterline/lib/waterline/utils/normalize.js:320:18)
        at callback.success (/usr/local/lib/node_modules/sails/node_modules/waterline/node_modules/switchback/lib/normalize.js:33:31)
        at _switch (/usr/local/lib/node_modules/sails/node_modules/waterline/node_modules/switchback/lib/factory.js:48:28)
        at /Users/shanegrey/Documents/Code/rmsApi/node_modules/sails-postgresql/lib/adapter.js:1159:16
        at __DROP__ (/Users/shanegrey/Documents/Code/rmsApi/node_modules/sails-postgresql/lib/adapter.js:303:13)
        at Query.handleReadyForQuery (/Users/shanegrey/Documents/Code/rmsApi/node_modules/sails-postgresql/node_modules/pg/lib/query.js:80:10)
        at null.<anonymous> (/Users/shanegrey/Documents/Code/rmsApi/node_modules/sails-postgresql/node_modules/pg/lib/client.js:158:19)
        at emitOne (events.js:82:20)
        at emit (events.js:169:7)

    Details:  AdapterError: Invalid connection name specified

    /usr/local/lib/node_modules/sails/lib/hooks/orm/index.js:221
                throw new Error(err);
                ^

    Error: [Error (E_UNKNOWN) Encountered an unexpected error] Details:  AdapterError: Invalid connection name specified

        at /usr/local/lib/node_modules/sails/lib/hooks/orm/index.js:221:19
        at taskCallback (/usr/local/lib/node_modules/sails/node_modules/async/lib/async.js:468:21)
        at /usr/local/lib/node_modules/sails/lib/hooks/orm/build-orm.js:57:23
        at /usr/local/lib/node_modules/sails/node_modules/waterline/lib/waterline.js:213:22
        at /usr/local/lib/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:183:21
        at /usr/local/lib/node_modules/sails/node_modules/waterline/lib/waterline/query/index.js:69:22
        at /usr/local/lib/node_modules/sails/node_modules/waterline/lib/waterline/adapter/sync/strategies/drop.js:36:10
        at wrapper (/usr/local/lib/node_modules/sails/node_modules/waterline/node_modules/lodash/index.js:3602:19)
        at applyInOriginalCtx (/usr/local/lib/node_modules/sails/node_modules/waterline/lib/waterline/utils/normalize.js:421:80)
        at wrappedCallback (/usr/local/lib/node_modules/sails/node_modules/waterline/lib/waterline/utils/normalize.js:331:16)
        at callback.error (/usr/local/lib/node_modules/sails/node_modules/waterline/node_modules/switchback/lib/normalize.js:42:31)
        at _switch (/usr/local/lib/node_modules/sails/node_modules/waterline/node_modules/switchback/lib/factory.js:46:28)
        at /usr/local/lib/node_modules/sails/node_modules/waterline/lib/waterline/adapter/ddl/index.js:44:22
        at wrapper (/usr/local/lib/node_modules/sails/node_modules/waterline/node_modules/lodash/index.js:3602:19)
        at applyInOriginalCtx (/usr/local/lib/node_modules/sails/node_modules/waterline/lib/waterline/utils/normalize.js:421:80)
        at wrappedCallback (/usr/local/lib/node_modules/sails/node_modules/waterline/lib/waterline/utils/normalize.js:331:16)

These are my connections:


    rms_dev: {
        adapter: 'sails-postgresql',
        host: 'localhost',
        user: 'rms',
        password: 'x',
        database: 'rms_dev'
    },

    rms_test: {
        adapter: 'sails-postgresql',
        host: 'rmsdbtest',
        user: 'rms',
        password: 'x',
        database: 'rms_test'
    },

    rms_prod: {
        adapter: 'sails-postgresql',
        host: 'rmsdbprod',
        user: 'rms',
        password: 'x',
        database: 'rms_prod'
    },

These are my versions:

    "sails": "~0.11.0",
    "sails-disk": "~0.10.0",
    "sails-hook-autoreload": "^0.11.5",
    "sails-hook-dev": "^1.0.0",
    "sails-postgresql": "^0.10.15",

**Comment by shanemgrey** [comment]
Which, noticing this is a mysql project, and the fact that I'm using postgres, I find myself thinking out that it's not a mysql connector issue only.

**Comment by devinivy** [comment]
I use the mysql adapter in production, and this has never been an issue.  So it seems like the cause might be around other software in the mix or configuration.  If you have a surefire way to reproduce, I'd be interested to experience the issue and find a fix.

**Comment by devinivy** [comment]
@shanemgrey what do you mean "if I change a controller and save"?  Do you have sails set to reload the ORM when files change?

**Comment by Ryanc1256** [comment]
I never could get it a surefire wag if reproducing it, it just would happen to show it self at the weirdest times.

**Comment by shanemgrey** [comment]
Yes, it *was* happening when I saved a controller file or model etc.  I was using "sails-hook-autoreload": "^0.11.5",

I'm not totally sure this is what fixed it because I couldn't figure out the exact series of events that triggered it.  But I've updated the autoreload and sails to the current versions.  And now I'm seeing the issue.

To be fair, I wasn't seeing the issue a week ago either, using the older versions of these files.  So it's likely there was some interaction with my code where I introduced a bug and was getting some obscured err in the trace.

In any event, I'm sorry I resolved the issue without being sure.

**Comment by sailsbot** [comment]
Thanks for posting, @geekhack.  I'm a repo bot-- nice to meet you!

The issue queue in this repo is for verified bugs with documented features.  Unfortunately, we can't leave some other types of issues marked as "open".  This includes bug reports about undocumented features or unofficial plugins, as well as feature requests, questions, and commentary about the core framework. Please review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) for details.

If you're here looking for help and can't find it, visit [StackOverflow](http://stackoverflow.com), our [Google Group](https://groups.google.com/forum/#!forum/sailsjs) or our [chat room](https://gitter.im/balderdashy/sails?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge).  But please do not let this disrupt future conversation in this issue! I am only marking it as "closed" to keep organized.

Thanks so much for your help!

> If I am mistaken, and this issue is about a _critical bug with a documented feature_, please accept my sincerest apologies-- I am but a humble bot working to make GitHub a better place.  If you open a new issue, a member of the core team will take a look ASAP.

**Comment by devinivy** [comment]
Definitely let us know if it crops up again!  Some interaction with sails-hook-autoreload was very likely the culprit, since that _will_ teardown connections.

**Comment by Ryanc1256** [comment]
@devinivy i would agree, but in our case we don't use sails-hook-autoreload or we don't use reload ORM hook either...

**Comment by devinivy** [comment]
@Ryanc1256 we can reopen this if you feel it's unresolved.  It might be worth updating to the latest version of waterline and adapters.  I'm also a little confused by the config you posted, since I could only find an adapter name, but not a connection name.

**Comment by Ryanc1256** [comment]
@devinivy yeah i have the lastest of waterline and the adapters, but the weird thing is it only seem's to happen on our live servers, so it's really hard to test what's actually breaking it, like that error hasn't come up in a couple of week's now, so I don't know if it's worth opening it or not.

**Comment by lucasfischer** [comment]
im getting the same error after updating node from 4.4 to 6.9.
Any ideas?
